### ISSUE TITLE
fig.colorbar messes up subplots centering

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#5463]
```
from pylab import *
fig, axs = plt.subplots(2)
axs[0].imshow(np.random.rand(32, 32), extent=(10, 0, 0, 10))
img = axs[1].imshow(np.random.rand(32, 32), extent=(10, 0, 0, 10))
fig.colorbar(img, ax=list(axs))
plt.show()
```

![colorbar](https://cloud.githubusercontent.com/assets/1322974/11099602/27f4af18-8861-11e5-8fdc-9f0971a22d9a.png)
(click on the image to confirm that it is indeed mis-centered)

**Comment by WeatherGod** [comment]
Confirmed. And it happens regardless if use_gridspec is True or False.

**Comment by mdboom** [comment]
I'm milestoning this as 2.0, but because I think we can guarantee we can fix this for then, but because fixing this may require changing behavior in a backward-incompatible way.

**Comment by anntzer** [comment]
Note to self and whoever hits this bug: this is due to the fact that axes with a set aspect ratio (directly or via an image) only positioned at drawing time (why?...) but the colorbar is positioned when it's drawn, so the colorbar thinks it's there for the normal, "rectangular" axes and everything collapes.

A stopgap measure is to call `fig.canvas.draw()` before creating the colorbar (because that'll resolve the positions of the axes).  In fact I wonder whether `plt.colorbar` should just do that... (at least in interactive mode, where that can't(?) be too bad except for a performance hit)

**Comment by efiring** [comment]
The automatic generation of the colorbar axes requires that it steal space from the main axes; setting the aspect ratio has to be done after this, hence at drawing time for the main axes.  This is the sort of chicken and egg problem that could be solved better with a constraints engine.  I don't think that triggering a full draw before creating the colorbar solves the problem; it just adds useless overhead.

**Comment by anntzer** [comment]
It's not perfect, but because triggering an initial draw brings the positions of the axes much closer to their final position, the final layout still looks much less unbalanced.

**Comment by efiring** [comment]
Sorry, I don't think the benefit is worth the cost.  It's just the wrong solution. At the stage of doing interactive plots, there is no functional advantage to improving the centering in this way.

**Comment by tacaswell** [comment]
Another possible solution is to use a function like:

```python
def squeeze_fig_aspect(fig, preserve='h'):
    preserve = preserve.lower()
    bb = bbase.union([ax.bbox for ax in fig.axes])

    w, h = fig.get_size_inches()
    if preserve == 'h':
        new_size = (h * bb.width / bb.height, h)
    elif preserve == 'w':
        new_size = (w, w * bb.height / bb.width)
    else:
        raise ValueError(
            'preserve must be "h" or "w", not {}'.format(preserve))
    fig.set_size_inches(new_size, forward=True)
```

after the first draw which re-sizes the figure to have the same aspect ratio as the bounding box around all of it's axes.

This probably needs some epicycles added like ignoring non-visible axes etc.

**Comment by anntzer** [comment]
Seems like a useful functionailty to have in general, but I'm not sure how this helps in this case?

**Comment by tacaswell** [comment]
And I am pretty sure that the fixed-aspect are positioned on the figure at draw time because the axes does not necessarily know the data-limits before then and the default behavior of a fixed aspect ratio axes is to resize the axes, not change the data limits, to get same ratio between data space and pixel space in x and y.

Subjectively, I think that the current behavior gives a better looking figure

```python
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.transforms import BboxBase as bbase

def squeeze_fig_aspect(fig, preserve='h'):
    preserve = preserve.lower()
    bb = bbase.union([ax.bbox for ax in fig.axes])

    w, h = fig.get_size_inches()
    if preserve == 'h':
        new_size = (h * bb.width / bb.height, h)
    elif preserve == 'w':
        new_size = (w, w * bb.height / bb.width)
    else:
        raise ValueError(
            'preserve must be "h" or "w", not {}'.format(preserve))
    fig.set_size_inches(new_size, forward=True)


fig, axs = plt.subplots(2)
axs[0].imshow(np.random.rand(32, 32), extent=(10, 0, 0, 10),
              cmap='viridis', interpolation='nearest')
img = axs[1].imshow(np.random.rand(32, 32), extent=(10, 0, 0, 10),
                    cmap='viridis', interpolation='nearest')

#fig.canvas.draw()
fig.colorbar(img, ax=list(axs))
squeeze_fig_aspect(fig)
```

![so](https://cloud.githubusercontent.com/assets/199813/12380335/9fa50410-bd3e-11e5-807b-ad3d450fb550.png)
![so2](https://cloud.githubusercontent.com/assets/199813/12380336/9fa6b3c8-bd3e-11e5-846f-ea6bac127277.png)

**Comment by anntzer** [comment]
I agree that the lower layout is much better (well, perhaps that raises other issues about automatic ticking...) but I don't see the figure being resized, either when calling the script directly or via `%run test.py` in IPython with %matplotlib mode on (so everything is still off-center).  Am I missing something?

**Comment by tacaswell** [comment]
Are you using `inline`?

**Comment by anntzer** [comment]
I wasn't.  In %inline mode the issue is not there to start with...

**Comment by jklymak** [comment]
TLDR: This is hard and needs someone to champion it (though basically the functionality is available with axesgrid).  

Just to explain what is going on here, and why it is hard to fix...

Suggest we forget about colorbars for a second, and consider 
```python
import numpy as np
import matplotlib.pyplot as plt

A = np.random.rand(24,8)
fig, axs = plt.subplots(2, 3, sharex=True, sharey=True, figsize=(6,3))
for ax in axs.flat:
    ax.pcolormesh(A)
    ax.set_aspect(1)
plt.show()
```

![all1](https://user-images.githubusercontent.com/1562854/47030051-bd9b5600-d121-11e8-843b-480747906728.png)

Matplotlib and its layout managers (tight_layout and constrained_layout) leave all the space between the axes because all the layout managers do is set the size of the grey boxes, not the size of the inner box that actually gets drawn. 

We could argue that a better layout would be to squeeze all the subplots together near the center and leave the blank space on the sides.  But that makes mixing subplots impossible:

![boo2](https://user-images.githubusercontent.com/1562854/47030135-f20f1200-d121-11e8-8be7-bf152420141e.png)

Should this case squeeze the upper row together or keep the axes above their partners in the lower row?  I'd argue the latter is the most expected.  

So, right now Matplotlib layout is done on the grey box level, and the same applies to colorbars.  They steal space from the grey box and are placed relative to it.  They don't know anything about the contents of the grey box.  Now, the colorbar code *also* yanks the subplot over next to the colorbar.  I think that behaviour can be changed, but it still does the easier thing, which is to place the colorbar relative to the grey box and tell the parent axes how to align the drawn axes inside the box.  

*Could* Matplotlib do this differently?  Absolutely.  There is already axesgrid, which builds a grid of the axes that is as compact as possible (given padding etc) in the direction that needs shrinking.  But it does so by assuming that all the subplots are of similar type.  I'm not sure what it would do w/ the sedond case above, but I don't think it would handle it.  

I've been considering moving axesgrid into the main library; I think it is an elegant solution to a problem substantial number of users would like solved (packing a bunch of fixed-aspect ratio plots together).  Because it is self-contained, it can make the oppositite assumption to tight_layout or constriained_layout and do the packing.  

My only hangup to working on this is that I have no personal need for such plots, or for the rare cases that I do, I just change the size of the figure so the white space doesn't look too bad.

**Comment by ZXP-S-works** [comment]
Try:
fig.colorbar(img, ax=list(axs), pad=0.4)

**Comment by jklymak** [comment]
The panchor kwarg also adresses this.

**Comment by jklymak** [comment]
I'm tempted to close this.  

```python
import matplotlib.pyplot as plt
import numpy as np
import matplotlib as mpl

fig, axs = plt.subplots(2, layout='compressed')
axs[0].imshow(np.random.rand(32, 32), extent=(10, 0, 0, 10))
img = axs[1].imshow(np.random.rand(32, 32), extent=(10, 0, 0, 10))
fig.colorbar(img, ax=list(axs))
plt.show()
```
gives
![cbarcenter](https://user-images.githubusercontent.com/1562854/226115071-053364f4-fc99-4089-901c-fd62170deecd.png)

If for some reason you don't want to use compressed layout, you can keep the axes centered with `panchor=(0.5, 0.5)` in the colorbar call.  This will orphan the colorbar to the right.  Or you can create the colorbar axes manually.

**Comment by anntzer** [comment]
Yup, compressed layout is great 👍

**Comment by benjaminpotter** [comment]
Glad this is resolved. Love compressed layout.
